cabby/rvs/templates.py

```diff
@@ -0,0 +1,164 @@
+AVOID = [
+        '. If you reach BEYOND_PIVOT, you have gone too far.',
+         '']
+GOAL_END = ['and meet at GOAL_LOCATION.']
+GOAL = ["GOAL_LOCATION"]
+
+
+def add_rules(nonterminal_name: Text, 
+list_terminals: Sequence[Text]) -> Sequence[Production]:
+    """Create the production rules for a givn nonterminal and a
+     list of terminals corresponding to it.
+
+    Args:
+      nonterminal_name: The name of the nonterminal.
+      list_terminals: The list of terminals that for each one a rule with 
+      the nonterminal will be produced.
+    Returns:
+      A sequence of productions rules.
+
+    """
+    prods = []
+    for phrase in list_terminals:
+        rule = Production(Nonterminal(nonterminal_name), (phrase,))
+        prods.append(rule)
+    return prods
+
+
+def create_templates():
+  """Creates the templates from the grammar."""
+
+  prods = [
+      Production(Nonterminal('S'), (Nonterminal('V2'), Nonterminal('V2_MAIN'))),
+      Production(Nonterminal('V2_MAIN'),
+                (Nonterminal('V1'), Nonterminal('M_G'))),
+
+      Production(Nonterminal('S'), (Nonterminal('V1'), Nonterminal('NO_GOAL'))),
+      Production(Nonterminal('S'), (Nonterminal(
+          'V1_GOAL'), Nonterminal('WITH_GOAL'))),
+      Production(Nonterminal('V1_GOAL'),
+                (Nonterminal('V1'), Nonterminal('V1_CON'))),
+      Production(Nonterminal('WITH_GOAL'),
+                (Nonterminal('GOAL'), Nonterminal('M_G'))),
+      Production(Nonterminal('M_G'),
+
+                (Nonterminal('MAIN_NO_V'), Nonterminal('E'))),
+      Production(Nonterminal('M_G'), (Nonterminal('MAIN'), Nonterminal('E'))),
+      Production(Nonterminal('E'), (Nonterminal(
+          'NEAR_GOAL_START'), Nonterminal('AVOID'))),    
+      Production(Nonterminal('E'), (Nonterminal(
+          'NEAR_GOAL_END'), Nonterminal('AVOID'))),
+      Production(Nonterminal('V1_CON'), ('to the',)),
+      Production(Nonterminal('NO_GOAL'),
+                (Nonterminal('MAIN_NO_V'), Nonterminal('G'))),
+      Production(Nonterminal('G'), (Nonterminal(
+          'GOAL_END'), Nonterminal('NEAR_GOAL_END'))),
+  ]
+
+  prods += add_rules('V2', V2)
+  prods += add_rules('AVOID', AVOID)
+  prods += add_rules('NEAR_GOAL_START', NEAR_GOAL_START)
+  prods += add_rules('NEAR_GOAL_END', NEAR_GOAL_END)
+  prods += add_rules('GOAL', GOAL)
+  prods += add_rules('GOAL_END', GOAL_END)
+  prods += add_rules('MAIN_NO_V', MAIN_NO_V)
+  prods += add_rules('MAIN', MAIN)
+  prods += add_rules('V1', V1)
+
+  grammar = CFG(Nonterminal('S'), prods)
+
+  # Generate templates.
+  templates = []
+  for sentence in generate(grammar):
+
+    sentence = ' '.join(sentence)
+
+    if sentence[-1] != '.':
+        sentence += '.'
+    sentence = sentence.replace(" .", ".")
+    sentence = sentence.replace(" ,", ",")
+    sentence = sentence.replace("..", ".")
+    templates.append(sentence)
+
+  templates_df = pd.DataFrame(templates, columns=['sentence'])
```

Comment:
Why use a data frame here instead of defining a data structure to hold these various elements and passing the templates around a sequence of Template items?

Reply:
easier to condition on.
E.g,
current_templates = gen_templates[
        current_templates['beyond_pivot'] == False]